### PR TITLE
MML-294 Fixing some issued with the UIAction and Dialog elements

### PR DESCRIPTION
**Pull request by symphony-mariacristina**
Goal of this PR is to fix some minor issues found during validation.
- Allowing the presence of whitespaces inside the Dialog element
- If a UIAction with a target-id referring to a Dialog comes in the messageML before the Dialog itself, the message should be allowed and parsed correctly.
- No more that one form is allowed inside a Dialog element
- We should be able to match Dialog-UIAction target id even if they are nested inside another element.
- Matching Dialog-UiAction should be in the same scope (at the same level of the messageML tree) to be valid.

Closes #295

### CODE REVIEW

File: src/test/java/org/symphonyoss/symphony/messageml/elements/UIActionTest.java [hunk 1]
```diff
@@ -345,15 +424,15 @@ public void testUIActionOpenDialogInDifferentScopeFromDialog() throws Exception
...
+            + "<form id=\"form-id\">"
             + "<ui-action trigger=\"click\" action=\"open-dialog\" target-id=\"target-dialog-id\">"
             + "<button>Open the dialog</button>"
             + "</ui-action>"
             + "</form>"
             + "</messageML>";
 
     expectedException.expect(InvalidInputException.class);
-    expectedException.expectMessage("ui-action with a target-id must have only one dialog sibling with a matching id");
+    expectedException.expectMessage("The form with id 'form-id' should have at least one action button");
```
Comment (AliciaSymphony) [inline src/test/java/org/symphonyoss/symphony/messageml/elements/UIActionTest.java:435]: Just to double check?
Does this appear every time that a form has an `id` ? Or only in the dialog context ?

File: src/main/java/org/symphonyoss/symphony/messageml/elements/Element.java [hunk 2]
```diff
@@ -634,6 +634,18 @@ void assertPhrasingContent() throws InvalidInputException {
...
 
+  /**
+   * Check that the element's children are limited to allowed element types returning a specific message given in input.
+   */
+  void assertContentModel(Collection<Class<? extends Element>> permittedChildren, String message)
+      throws InvalidInputException {
+    try {
+      assertContentModel(permittedChildren);
+    } catch (InvalidInputException e) {
+      throw new InvalidInputException(message);
```
Comment (symphony-youri) [inline src/main/java/org/symphonyoss/symphony/messageml/elements/Element.java:645]: do we hide the original exception on purpose here? vs chaining it with throw new InvalidInputException(message, e) ?

File: src/main/java/org/symphonyoss/symphony/messageml/elements/Element.java [hunk 3]
```diff
@@ -770,6 +782,23 @@ void assertChildrenNotExceedingMaxCount(Collection<Class<? extends Element>> ele
     }
   }
 
+  /**
+   * Given a target id present in a UIAction this method returns the corresponding dialog associated to the same
+   * id, if no dialog is found it throws an exception
+   */
+  Dialog checkMatchingDialog(String targetId) throws InvalidInputException {
```
Comment (symphony-youri) [inline src/main/java/org/symphonyoss/symphony/messageml/elements/Element.java:789]: checkMatching -> findMachingDialog? (as we don't really check the dialog itself but just look for one)

File: src/main/java/org/symphonyoss/symphony/messageml/elements/MessageML.java [hunk 4]
```diff
@@ -155,4 +161,19 @@ public void setChime(boolean chime) {
   public String getPresentationMLTag() {
     return PRESENTATIONML_TAG;
   }
+
+  /**
+   * If the messageML contains a uiAction with a target-id this method checks that exists a corresponding
+   * dialog element with the same id.
+   */
+  private void validateTargetIdForUIActions() throws InvalidInputException {
+    List<Element> uiActionWithTargetId = getChildren().stream()
```
Comment (symphony-youri) [inline src/main/java/org/symphonyoss/symphony/messageml/elements/MessageML.java:170]: can it be nested several level down? (like a ui action inside a container? like a form or a div?)

Comment (symphony-mariacristina) [reply src/main/java/org/symphonyoss/symphony/messageml/elements/MessageML.java:170]: you are right, we should search also for nested UIActions.

File: src/main/java/org/symphonyoss/symphony/messageml/elements/Element.java (hunk 2 above)
Comment (symphony-mariacristina) [reply src/main/java/org/symphonyoss/symphony/messageml/elements/Element.java:645]: I wanted to make it more specific that's why I'm completely replacing the original message. 
It was too generic for this usecase.

File: src/test/java/org/symphonyoss/symphony/messageml/elements/UIActionTest.java (hunk 1 above)
Comment (symphony-mariacristina) [reply src/test/java/org/symphonyoss/symphony/messageml/elements/UIActionTest.java]: There was actually a mistake in the test, I added the button to the form and now we are back to the first message error :)

**Comment by symphony-mariacristina** [comment]
I'm wondering if this post processing is a bit too heavy for each message. A possible improvement could be to add a flag to the messageML to be set when a UIAction with target-id is found during messageML parsing. In this case we only do this post processing when we know for sure that there's a match to be found. 
We already do something similar for chimes https://github.com/symphonyoss/messageml-utils/blob/master/src/main/java/org/symphonyoss/symphony/messageml/elements/MessageML.java#L130
Any thoughts @symphony-youri @symphony-thibault ?

**Comment by symphony-youri** [comment]
> I'm wondering if this post processing is a bit too heavy for each message. A possible improvement could be to add a flag to the messageML to be set when a UIAction with target-id is found during messageML parsing. In this case we only do this post processing when we know for sure that there's a match to be found.
> We already do something similar for chimes https://github.com/symphonyoss/messageml-utils/blob/master/src/main/java/org/symphonyoss/symphony/messageml/elements/MessageML.java#L130
> Any thoughts @symphony-youri @symphony-thibault ?

This is just about navigating a few elements in the tree right? I would say usually messages don't have many elements (like hundreds max?) so it should be pretty fast.
But if you want to confirm that you can use JMH to assess the impact (with/without post processing) (MessageMLContextBenchmark)

**Comment by symphony-mariacristina** [comment]
I did a test using the JMH with a complex message and seems like the post processing is not impacting the performance of the MessageML parsing, so I guess we can keep it as it is.

**Comment by symphony-mariacristina** [comment]
> do you get different results now?

Not so much fortunately :)

Results with the post processing:
```
Benchmark                                                     Mode  Cnt    Score    Error  Units
MessageMLContextBenchmark.parseComplexMessageMLWithEntities  thrpt   25  826.608 ± 19.860  ops/s
```
Results without:
```
Benchmark                                                     Mode  Cnt    Score    Error  Units
MessageMLContextBenchmark.parseComplexMessageMLWithEntities  thrpt   25  822.082 ± 20.188  ops/s
```

**Review by symphony-youri — APPROVED** [review]

**Pull request merged**